### PR TITLE
feat(tabs): animate active tab underline

### LINKED ISSUE #867: Enhancement: Tabs — add slide transition animation to active bar
Please add slide transition animation to active bar. ease-in-out. Preferably variable speed based on distance (e.g. if one tab away, .2s; if 5 tabs away .4s)

image

### PR DESCRIPTION
**Pull request by paulcpederson**
Related Issue: Resolves #867 

## Summary

This resolves the last couple issues in @macandcheese 's tab style pr regarding console warnings on update. I couldn't push to his branch as it's on a fork so I'm creating a new pr with his work + this one small update. Once this is in we will be able to close #893

I've also updated this branch with master so it will merge cleanly.

### CODE REVIEW

File: src/components/calcite-tab-nav/calcite-tab-nav.tsx [hunk 1]
```diff
@@ -199,18 +266,63 @@ export class CalciteTabNav {
...
+    this.indicatorWidth = this.selectedTabEl?.offsetWidth;
+  }
+
   private getIndexOfTabTitle(el: HTMLCalciteTabTitleElement, tabTitles = this.tabTitles): number {
     // In most cases, since these indexes correlate with tab contents, we want to consider all tab titles.
     // However, when doing relative index operations, it makes sense to pass in this.enabledTabTitles as the 2nd arg.
     return tabTitles.indexOf(el);
   }
 
+  private getTabTitleById(id: string | number): Promise<HTMLCalciteTabTitleElement | null> {
```
Comment (driskull) [inline src/components/calcite-tab-nav/calcite-tab-nav.tsx:320]: nitpick: could use async await here :)

File: src/components/calcite-tab-nav/calcite-tab-nav.tsx [hunk 2]
```diff
@@ -199,18 +266,63 @@ export class CalciteTabNav {
...
+  //
+  //--------------------------------------------------------------------------
+
+  //--------------------------------------------------------------------------
+  //
+  //  Private State/Props
+  //
+  //--------------------------------------------------------------------------
+
+  private tabNavEl: HTMLElement;
```
Comment (driskull) [inline src/components/calcite-tab-nav/calcite-tab-nav.tsx:281]: question: do we know the element types here or could it be any HTMLElement?

File: src/components/calcite-tab-nav/calcite-tab-nav.tsx [hunk 3]
```diff
@@ -116,12 +132,55 @@ export class CalciteTabNav {
...
+          }
+        : {
+            width: `${this.indicatorWidth}px`,
+            right: `${this.indicatorOffset}px`
+          };
+    return (
+      <Host role="tablist">
+        <div
+          class="tab-nav"
+          onScroll={() => this.handleContainerScroll()}
```
Comment (driskull) [inline src/components/calcite-tab-nav/calcite-tab-nav.tsx:152]: nitpick: I think you could do `private handleContainerScroll = (): void => {` and then just `onScroll={this.handleContainerScroll}`

File: src/components/calcite-tab-nav/calcite-tab-nav.tsx [hunk 4]
```diff
@@ -116,12 +132,55 @@ export class CalciteTabNav {
     }
   }
 
+  render(): VNode {
+    this.dir = getElementDir(this.el);
+
```
Comment (driskull) [inline src/components/calcite-tab-nav/calcite-tab-nav.tsx:137]: nitpick:

```
const width = `${this.indicatorWidth}px`;
const offset = `${this.indicatorOffset}px`;

...


{
            width,
            left: offset
          }

```

**Review by driskull — APPROVED** [review]
Looks good! just nitpicky suggestions :D

File: src/components/calcite-tab-nav/calcite-tab-nav.tsx [hunk 5]
```diff
@@ -199,18 +266,63 @@ export class CalciteTabNav {
...
   //
   //--------------------------------------------------------------------------
 
+  private handleContainerScroll(): void {
+    // remove active indicator transition duration while container is scrolling to prevent wobble
+    this.activeIndicatorEl.style.transitionDuration = "0s";
+    this.getOffsetPosition();
+  }
+
+  private getOffsetPosition(): void {
```
Comment (jcfranco) [inline src/components/calcite-tab-nav/calcite-tab-nav.tsx:303]: Can you rename this? It's conveying a getter, but not returning anything.

File: src/components/calcite-tab-nav/calcite-tab-nav.tsx [hunk 6]
```diff
@@ -199,18 +266,63 @@ export class CalciteTabNav {
...
+  //  Private State/Props
+  //
+  //--------------------------------------------------------------------------
+
+  private tabNavEl: HTMLElement;
+
+  private activeIndicatorEl: HTMLElement;
+
+  // the duration of active indicator animation
+  private animationActiveDuration = "0.3s";
```
Comment (jcfranco) [inline src/components/calcite-tab-nav/calcite-tab-nav.tsx:286]: Nitpick: can this be renamed to convey that it's a CSS value and not a number?

File: src/components/calcite-tab-nav/calcite-tab-nav.tsx (hunk 2 above)
Comment (paulcpederson) [reply src/components/calcite-tab-nav/calcite-tab-nav.tsx:281]: I think we could type this more specifically to the `HTMLCalciteTabNavElement`, will try it out

File: src/components/calcite-tab-title/calcite-tab-title.tsx [hunk 7]
```diff
@@ -197,7 +197,7 @@ export class CalciteTabTitle {
   /**
    * @internal
    */
-  @Event() calciteTabTitleRegister: EventEmitter;
+  @Event() calciteTabTitleRegister: EventEmitter<string | number>;
```
[the remark this answers is not in the document]
Comment (paulcpederson) [reply src/components/calcite-tab-title/calcite-tab-title.tsx:200]: apparently the tab signifier is sometimes the index of the tab in it's parent? IDK, I found it kind of confusing but preserved the type that's used elsewhere.

**Comment by paulcpederson** [comment]
thanks for review both, have a couple other things to wrap up in another branch and then I'll push in updates with this feedback

**Comment by paulcpederson** [comment]
OK all feedback should now be addressed.

File: src/components/calcite-tab-title/calcite-tab-title.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulcpederson) [reply src/components/calcite-tab-title/calcite-tab-title.tsx]: Made a type for this at least

**Pull request merged**